engine/src/main/java/io/zeebe/engine/processing/common/CatchEventBehavior.java

```diff
@@ -155,7 +166,13 @@ public void subscribeToTimerEvent(
         .setTargetElementId(handlerNodeId)
         .setProcessDefinitionKey(processDefinitionKey);
 
-    commandWriter.appendNewCommand(TimerIntent.CREATE, timerRecord);
+    sideEffects.add(
+        () -> {
+          timerChecker.scheduleTimer(dueDate);
+          return true;
+        });
```

Comment:
I thought this should happen in the TimerCreatedApplier. Timer's are not state, but also kinda are. After a restart (and replay), we should have a timer set for each of the timers that were created. Can you explain why you placed it here? Also, I'm not sure if we have one, but we might need a test to check that timers are created and triggered after replay.

Reply:
One thing to remember here is the cardinality between scheduled timer tasks and timers in the state. There is at most one scheduled timer task at any time. If no task is scheduled, and `scheduletimer()` is called, we schedule that one task. If a task is already scheduled, then 'scheduleTimer()` simply checks whether the new due date is before or after the currently scheduled due date. If it is, the scheduled task is cancelled and rescheduled with the earlier due date. Otherwise, it does nothing. Then, whenever a scheduled task is executed, it will look at the state to schedule the next timer task. That's why we ever only have one, which is the one for the next invocation. So in that sense, the state in the database weighs heavier than the one scheduled timer task.

Doing this bit in the event applier has one problem: If the due date (which is stored in the event) is in the past, then we cannot schedule a timer task for a past time. The right thing to do here would be to send out a `TimerIntent.TRIGGER` command, which we cannot do in an event applier. So that is why I think the event applier is not the right place to put it.

This, of course doesn't answer the question "what happens on replay". In fact, nothing happens during replay, but after replay `iDueDateTimerChecker#onRecovered()` is called, which first looks at the state to find all timers which would have fired in the past and triggers them immediately, and then schedules the next timer. This part of the code base hasn't changed.

Regarding the tests, I am not sure. We will definitely have coverage of this when the other PR regarding timers in randomized workflows is merged. Then we will create timers, and we will stop and restart the engine. If timers were not rescheduled on restart, then these tests should fail.

Reply:
Thanks for the comprehensive explanation. Perhaps we can add a comment above line 169 that says something like: `timerChecker implements onRecovered to recover from restart, so no need to schedule this in TimerCreatedApplier`.